mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,162 @@
+    mut curr_id: Id<Block>,
+    stop_id: Id<GenBlock>,
+    mut func: impl FnMut(Block),
+) -> Result<(), ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    while curr_id != stop_id {
+        let block = chainstate
+            .get_block(curr_id)?
+            .ok_or_else(|| ReorgError::BlockNotFound(curr_id))?;
+        curr_id = match block.prev_block_id().classify(chain_config) {
+            common::chain::GenBlockId::Genesis(_) => break,
+            common::chain::GenBlockId::Block(block_id) => block_id,
+        };
+        func(block);
+    }
+    Ok(())
+}
+
+fn find_disconnected_txs(
+    chainstate: &dyn ChainstateInterface,
+    old_tip_id: Id<GenBlock>,
+    new_tip_id: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let old_tip_id = match old_tip_id.classify(chain_config) {
+        common::chain::GenBlockId::Genesis(_) => return Ok(Vec::new()),
+        common::chain::GenBlockId::Block(id) => id,
+    };
+
+    let common_id = {
+        let old_index = chainstate.get_block_index(&old_tip_id)?.ok_or(ReorgError::OldTipIndex)?;
+        let new_index = chainstate.get_block_index(&new_tip_id)?.ok_or(ReorgError::NewTipIndex)?;
+        let common_index = chainstate.last_common_ancestor(&old_index.into(), &new_index.into())?;
+        common_index.block_id()
+    };
+
+    // Short circuit the expensive processing below if no blocks are being disconnected
+    if old_tip_id == common_id {
+        return Ok(Vec::new());
+    }
+
+    // Collect IDs of trancsactions included in the newly connected chain
+    let mut connected_txs = BTreeSet::new();
+    for_each_block(chainstate, new_tip_id, common_id, |block| {
+        connected_txs.extend(block.transactions().iter().map(|t| t.transaction().get_id()))
+    })?;
+
+    // Collect txns that have been removed from the blockchain and not added on the new fork
+    let mut disconnected_txs = Vec::new();
+    for_each_block(chainstate, old_tip_id, common_id, |block| {
+        // We iterate blocks in the reverse order, so for consistent transaction input/output
+        // dependencies, take transacitons in reverse too,
+        let txns = block.transactions().iter().rev();
+        // We disregerd transactions that are re-added on the newly connected chain
+        let txns = txns.filter(|txn| !connected_txs.contains(&txn.transaction().get_id()));
+        disconnected_txs.extend(txns.cloned())
+    })?;
+
+    Ok(disconnected_txs)
+}
+
+fn fetch_disconnected_txs<M>(
+    mempool: &Mempool<M>,
+    new_tip: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chainstate = mempool.blocking_chainstate_handle();
+
+    let old_tip = mempool.tx_verifier.get_best_block_for_utxos().map_err(|_| ReorgError::OldTip)?;
+    if old_tip == new_tip {
+        return Ok(Vec::new());
+    }
+
+    chainstate.call(move |c| find_disconnected_txs(c.as_ref(), old_tip, new_tip))?
```

Comment:
Isn't it better to reduce the load on chainstate by loading all the relevant blocks to the mempool thread and then process them one by one?

Reply:
Done